### ISSUE TITLE
Support for having the IPFS network inside I2P

### ISSUE THREAD

**Issue by TheNain38** [ipfs/notes#124]
I2P is peer-to-peer friendly, and is distributed, which Tor isn't...
By using the [SAM API](https://geti2p.net/en/docs/api/samv3)
There are two library written in Go:
- [sam3](https://bitbucket.org/kallevedin/sam3) is complete
- [goSAM](https://github.com/cryptix/goSam) which is missing support for datagrams (Repliable and Raw). So, you can't really use it for IPFS which has an UDP based transport.

**Comment by jbenet** [comment]
Would be great to have I2P as a libp2p transport, in both Go and JS! \o/

Thanks for the lib links.

There is also this new Go implementation of I2P router: https://github.com/hkparker/go-i2p cc @hkparker
 (im not familiar enough with I2P yet to know how this all fits together).

**Comment by jbenet** [comment]
cc @cryptix too as i think he's interested in I2P.

**Comment by hkparker** [comment]
I'm just getting started with my implementation so its not on the network yet, but I'd be happy to help anyone who wants to contribute at this point!

**Comment by cryptix** [comment]
Thanks for the notification @jbenet!

It's quite the undertaking, thanks to @hkparker! You have seen the (wip) c++ [i2pd](https://github.com/PurpleI2P/i2pd) implementation, right? I'm sure you could exchange a lot with those guise. 

btw @TheNain38: I'm happy to add datagrams to my sam lib, it's just that nobody asked for it and I don't need it (yet).

**Comment by str4d** [comment]
@cryptix I haven't checked if your SAM lib supports v3.2, but it would be useful to do so, because that adds support for ports inside I2P. Also, v3.3 enables running streaming and datagrams over the same Destination, which could be handy :smile: 

Also, hi! I'm an I2P dev, happy to answer questions.

**Comment by majestrate** [comment]
I have a (toy) implementation of the go `net` package that provides `net.Listener` `net.Conn` `net.Addr` `net.PacketConn` and friends that use i2p. would need to change where the repo is located however, it's stuck in my misc i2p tools repo [here](https://github.com/majestrate/i2p-tools/tree/master/lib/i2p), not sure what would be needed for ipfs however.

**Comment by jbenet** [comment]
@majestrate may be good to extract that out into own package?

**Comment by majestrate** [comment]
@jbenet yes I plan on extracting it to its own package/repo sometime soon (possibly today)

**Comment by Zogg** [comment]
Any progress on this?

**Comment by cryptix** [comment]
> Any progress on this?

with the recent work on tor onion support (ipfs/notes#37) this should be easier.

**Comment by DaniellMesquita** [comment]
Adding reference data to this issue for better understanding.
This issue has been started on 2016 by @TheNain38, but just on 2018 the IPFS team (@jbenet and @daviddias) added labels to it.
@hkparker has been mentioned by jbenet as being the creator of a [new Go implementation of i2P router](https://github.com/hkparker/go-i2p). He told that wants help from users on improving it. I ask if it is the same as https://bitbucket.org/kallevedin/sam3 or @cryptix 's  [SAM](https://geti2p.net/en/docs/api/samv3).

> It's quite the undertaking, thanks to @hkparker! You have seen the (wip) c++ i2pd implementation, right? I'm sure you could exchange a lot with those guise.

Is that already being used?

@str4d
I remember you. You'd was helping to implement i2P through ZeroNet.

Before this issue, on 2015 users were already asking about implementing i2P on IPFS (@aaannndddyyy asked on https://github.com/ipfs/faq/issues/66)

@whyrusleeping @geebotron

**Comment by agowa338** [comment]
I've seen i2P and Tor being mentioned within the [libp2p README.md](https://github.com/libp2p/libp2p/blob/master/README.md).
So is there an update on this issue, too?

**Comment by marten-seemann** [comment]
> I've seen i2P and Tor being mentioned within the libp2p README.md.

Where do you see that? There's no mention of either i2p or Tor in that file. It's mentioned in the Visionary section of the [roadmap](https://github.com/libp2p/specs/blob/master/ROADMAP.md), but it's not something we'll be working on any time soon.

**Comment by agowa338** [comment]
![image](https://user-images.githubusercontent.com/2544867/129105228-fa63d57c-217c-471d-974e-35a1e797b341.png)

**Comment by marten-seemann** [comment]
Most of those protocols are actually not implemented (yet). Again, this is not something that has priority at the moment.

**Comment by agowa338** [comment]
Ok, thought the graphics show the current state of the protocol stack.
